Default resources

pdamer opened the pull request:
This is a start.  There is a small bug that the resource edit form sometimes pops up with default checked even though the resource isn't set as a default.  I wasn't able to test the code that should automatically add a user to all default resources so we can take that out if you want.

pdamer commented:
This fixes #49

pacna commented on database/resource.go:
```diff
@@ -68,6 +71,10 @@ type Resource struct {
+    // Default state of the Resource
+    // required: true
+    // example: true
+    IsDefault bool `json:"is_default"`
```
Would it be alright to change `is_default` to `isDefault`? All the data model properties in the UI are camelCase.

pdamer replied:
sure thing

pacna commented on database/resource.go:
```diff
@@ -92,6 +99,10 @@ type ResourceRequest struct {
+    // Default state of the Resource
+    // required: true
+    // example: true
+    IsDefault bool `json:"is_default"`
```
Same

pacna commented on database/resource.go:
```diff
@@ -104,6 +115,10 @@ type RegisterResourceRequest struct {
+    // Default state of the Resource
+    // required: false
+    // example: true
+    IsDefault bool `json:"is_default"`
```
Same

pacna commented on ui/js/service/resource.service.ts:
```diff
@@ -37,13 +37,15 @@ export class ResourceService extends HTTPService {
-    address: string;
     name: string;
+    address: string;
+    is_default: boolean;
```
Please change this to `isDefault` for consistency

pacna commented on ui/js/service/resource.service.ts:
```diff
@@ -37,13 +37,15 @@ export class ResourceService extends HTTPService {
     id: string;
     name: string;
+    address: string;
+    is_default: boolean;
```
Same

pacna commented on ui/js/service/resource.service.ts:
```diff
@@ -61,6 +63,7 @@ export namespace ResourceService {
     address: string;
     id: string;
     name: string;
+    is_default: boolean;
```
Same

dfirebaugh commented on database/resource.go:
```diff
@@ -226,6 +242,32 @@ func (db *Database) AddUserToResource(email string, resourceID string) (MemberRe
 	return memberResource, nil
 }
 
+// AddUserToDefaultResources - grants a user access to default resources - untested
+func (db *Database) AddUserToDefaultResources(email string) ([]MemberResourceRelation, error) {
+
+	m, err := db.GetMemberByEmail(email)
+	if err != nil {
+        //is this wrong?
+		return []MemberResourceRelation{}, err
+	}
+
+	rows, err := db.pool.Query(context.Background(), insertMemberDefaultResourceQuery, m.ID)
+	if err != nil {
+		log.Fatalf("conn.Query failed: %v", err)
+	}
+
+	defer rows.Close()
+
+	var memberResources []MemberResourceRelation
+
+	for rows.Next() {
+		var r MemberResourceRelation
+		err = rows.Scan(&r.ID, &r.MemberID, &r.ResourceID)
+		memberResources = append(memberResources, r)
+	}
+	return memberResources, nil
+}
```
This is fine

```suggestion
// AddUserToDefaultResources - grants a user access to default resources - untested
func (db *Database) AddUserToDefaultResources(email string) ([]MemberResourceRelation, error) {

	m, err := db.GetMemberByEmail(email)
	if err != nil {
		return []MemberResourceRelation{}, err
	}

	rows, err := db.pool.Query(context.Background(), insertMemberDefaultResourceQuery, m.ID)
	if err != nil {
		log.Fatalf("conn.Query failed: %v", err)
	}

	defer rows.Close()

	var memberResources []MemberResourceRelation

	for rows.Next() {
		var r MemberResourceRelation
		err = rows.Scan(&r.ID, &r.MemberID, &r.ResourceID)
		memberResources = append(memberResources, r)
	}
	return memberResources, nil
}
```

alternatively, we could move the declaration up
```suggestion
// AddUserToDefaultResources - grants a user access to default resources - untested
func (db *Database) AddUserToDefaultResources(email string) ([]MemberResourceRelation, error) {
	var memberResources []MemberResourceRelation

	m, err := db.GetMemberByEmail(email)
	if err != nil {
		return memberResources, err
	}

	rows, err := db.pool.Query(context.Background(), insertMemberDefaultResourceQuery, m.ID)
	if err != nil {
		log.Fatalf("conn.Query failed: %v", err)
	}

	defer rows.Close()


	for rows.Next() {
		var r MemberResourceRelation
		err = rows.Scan(&r.ID, &r.MemberID, &r.ResourceID)
		memberResources = append(memberResources, r)
	}
	return memberResources, nil
}
```

dfirebaugh commented on payments/payments.go:
```diff
@@ -71,6 +71,8 @@ func mapMemberIDToPayments(payments []database.Payment) []database.Payment {
 			if err != nil {
 				log.Errorf("error adding member to DB: %s", err.Error())
 			}
+            //TODO: test that this adds the member to all default resources.
```
```suggestion
```

this seems to work with the query update i suggested in my other comment

dfirebaugh commented:
When I was testing this out, i found a weird issue in the scheduler.
Seems like that just loops for a while tryiing to subscribe and publish to the MQTT broker.

I'll comment it out of the main branch and open an issue.

Just wanted you to be aware of it in case you run into that too.

pdamer commented:
Used @dfirebaugh's query. Switched over to isDefault and incorporated refactor updates.  Also figured out the problem with the checkbox.  Needed a ? in front.  Should be ready to go.
